Add support for logging additional request headers

dustinsoftware opened the pull request:
# Add support for logging additional request headers

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Summary of the changes (Less than 80 chars)

## Description

Adds an _optional_ AdditionalRequestHeaders field to W3CLoggerOptions, which accepts a string list of request headers to be logged.

I'm unsure what docs need to be updated, but happy to submit a patch for those too. 

Fixes #41698

Commit:
Add support for logging additional request headers

wtgodbe commented on src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs:
```diff
@@ -181,6 +183,17 @@ public async Task Invoke(HttpContext context)
+                    {
+                        if (headers.TryGetValue(_options.CurrentValue.AdditionalRequestHeaders[i], out var headerValue))
+                        {
+                            shouldLog |= AddToList(elements, i + _fieldsLength, headerValue.ToString());
```
What if the user passes an invalid header name? I think right now that invalid header name would wind up in the fields directive, then would be logged with a `-` on every line? My inclination is we should throw in that case, though I'm curious what @Tratcher @BrennanConroy @halter73 think

Tratcher replied:
Is that different from if they specified a header that the client never sends? I wouldn't worry about it.

wtgodbe commented on src/Middleware/HttpLogging/test/W3CLoggingMiddlewareTests.cs:
```diff
@@ -103,6 +103,48 @@ public async Task DefaultDoesNotLogOptionalFields()
         Assert.DoesNotContain(lines[3], "Snickerdoodle");
     }
 
+    [Fact]
```
Add a test for if the user passes an invalid header name (once we decide on a behavior for that)

wtgodbe replied:
Would still like a test that confirms that passing an invalid header doesn't cause the middleware to throw, and that the log file looks how we expect

dustinsoftware replied:
Added here a792f7ece52e774bc84b69f59f36bb3702a675d6

wtgodbe commented on src/Middleware/HttpLogging/src/W3CLoggerOptions.cs:
```diff
@@ -106,6 +106,11 @@ public TimeSpan FlushInterval
         }
     }
 
+    /// <summary>
```
Add a warning about potential security concerns here, similar to the comment on Cookie in `W3CLoggingFields`: https://github.com/dotnet/aspnetcore/blob/bf50999cd8e60d4c683b772dffa9898794a8433a/src/Middleware/HttpLogging/src/W3CLoggingFields.cs#L112-L120

wtgodbe replied:
CC @blowdart

dustinsoftware replied:
added

wtgodbe commented on src/Middleware/HttpLogging/samples/Logging.W3C.Sample/Startup.cs:
```diff
@@ -15,6 +15,10 @@ public void ConfigureServices(IServiceCollection services)
         {
             // Log all W3C fields
             logging.LoggingFields = W3CLoggingFields.All;
+            logging.AdditionalRequestHeaders = new[] {
```
Could you share what the log file looks like after this change to the sample?

dustinsoftware replied:
```
#Version: 1.0
#Start-Date: 2022-05-20 18:31:34
#Fields: date time c-ip cs-username s-computername s-ip s-port cs-method cs-uri-stem cs-uri-query sc-status time-taken cs-version cs-host cs(User-Agent) cs(Cookie) cs(Referer) cs(x-forwarded-for) cs(x-client-ssl-protocol)
2022-05-20 18:31:34 127.0.0.1 - VVKTV23H53 127.0.0.1 5000 GET / - 200 19.767 HTTP/1.1 localhost:5000 curl/7.79.1 - - 1.1.1.1 TLS+1.2
```

wtgodbe reviewed (changes requested):
Looks mostly good, with a few small caveats. Thanks!

wtgodbe commented on src/Middleware/HttpLogging/src/W3CLoggerOptions.cs:
```diff
@@ -106,6 +106,16 @@ public TimeSpan FlushInterval
     }
 
+    /// <summary>
+    /// List of additional request header values to log.
```
Update this to talk about request headers, not just cookies (e.g. `Request Headers can contain private information which may have... Arbitrary Request Headers should not be logged unless...`

halter73 replied:
Whatever general warnings we add about the security concerns of logging request headers here we should also add to `HttpLoggingOptions`.

dustinsoftware replied:
added to both places

halter73 commented on src/Middleware/HttpLogging/src/W3CLoggerOptions.cs:
```diff
@@ -106,6 +106,16 @@ public TimeSpan FlushInterval
+    /// unless logs are secure and access controlled
+    /// and the privacy impact assessed.
+    /// </summary>
+    public IReadOnlyList<string>? AdditionalRequestHeaders { get; set; }
```
Thanks for doing this @dustinsoftware!

#41698 is still labeled `api-ready-for-review` which means that the public API changes have not been reviewed and are likely to change. There's even a chance the new API will be rejected entirely. I don't think that's very likely in this particular case, but it's something to be aware of when working on new API proposals that don't have the `api-approved` label yet.

I think this should probably follow the pattern set forth in [HttpLoggingOptions.RequestHeaders](https://github.com/dotnet/aspnetcore/blob/95374b37562dc97290f44b14f15196784550b8d1/src/Middleware/HttpLogging/src/HttpLoggingOptions.cs#L18-L25) and use a get-only `ISet<string>` for these headers unless there's a problem with that approach.

```suggestion
    public ISet<string>? AdditionalRequestHeaders { get; }
```

Would it also make sense to also add a `ResponseHeaders` property to `W3CLoggerOptions` like we have in `HttpLoggingOptions`? Or would it be hard to distinguish request and response headers because of the W3C logging format?

dustinsoftware replied:
> Would it also make sense to also add a ResponseHeaders property to W3CLoggerOptions like we have in HttpLoggingOptions? Or would it be hard to distinguish request and response headers because of the W3C logging format?

Yeah I think that makes sense although that wasn't proposed in the original issue. To avoid this one getting blocked, I'd like to open a separate issue after this PR lands, unless they both need to be merged at the same time.

dustinsoftware replied:
> I think this should probably follow the pattern set forth in [HttpLoggingOptions.RequestHeaders](https://github.com/dotnet/aspnetcore/blob/95374b37562dc97290f44b14f15196784550b8d1/src/Middleware/HttpLogging/src/HttpLoggingOptions.cs#L18-L25) and use a get-only ISet<string> for these headers unless there's a problem with that approach.

I didn't do that here since there is no ctor with args for W3CLoggingOptions, so the caller is relying on the public setter for the instance passed from the AddW3CLogging callback:

```
        services.AddW3CLogging(logging =>
        {
            // Log all W3C fields
            logging.LoggingFields = W3CLoggingFields.All;
            logging.AdditionalRequestHeaders = new[] {
                "x-forwarded-for",
                "x-client-ssl-protocol",
            }.ToHashSet();
        });
```

halter73 replied:
> Yeah I think that makes sense although that wasn't proposed in the original issue.

Makes sense. We don't expect you to add any public API other than the approved `public ISet<string> AdditionalRequestHeaders { get; }`. I did mention this in the API review notes as a possible *future* enhancement.

> 2. Can we make this option more generic to support more field types like response headers?
> Maybe if we parsed prefixes like cs and sc, but that seems complicated. We could add a ResponseHeaders option in the future if we want.

`HttpLoggingOptions` doesn't have a constructor either. We expect the developer to use `.Clear()`/`.Add(string)`/`.Remove(string)` on the existing `ISet<string>` object to configure it:

https://github.com/dotnet/aspnetcore/blob/95374b37562dc97290f44b14f15196784550b8d1/src/Middleware/HttpLogging/test/HttpLoggingOptionsTests.cs#L57-L60

dustinsoftware replied:
I see. Updated :)

halter73 commented on src/Middleware/HttpLogging/src/W3CLogger.cs:
```diff
@@ -47,7 +47,8 @@ private string Format(string[] elements)
         {
-            if (_loggingFields.HasFlag((W3CLoggingFields)(1 << i)))
+            // Custom headers start at index 17
+            if (_loggingFields.HasFlag((W3CLoggingFields)(1 << i)) || i >= 17)
```
`|| i >= 17` looks super fragile. We should just pass `string[] additionalHeaders` as a separate array. (e.g. `_w3cLogger.Log(elements, additionalHeaders)`. Or, if that's too inefficient (which I don't think it is relative to everything else), we could pass the index `int startHeaderIndex` (e.g. `_w3cLogger.Log(elements, _fieldsLength)`).

---

FWIW, in `HttpLoggingOptions`, we don't log any of the request headers specified in the `ISet<string>` unless `HttpLoggingFields.RequestHeaders` is set.

https://github.com/dotnet/aspnetcore/blob/8a4b4deb09c04134f22f8d39aae21d212282004f/src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs#L101-L104

It doesn't make much sense to follow that pattern here since `RequestHeaders` isn't its own flag like it is in `HttpLoggingFields` and instead is the combination of the flags for specific request headers.
 https://github.com/dotnet/aspnetcore/blob/8a4b4deb09c04134f22f8d39aae21d212282004f/src/Middleware/HttpLogging/src/W3CLoggingFields.cs#L139
 
And the default value for `W3CLoggerOptions.AdditionalRequestHeaders` is empty unlike `HttpLoggingOptions.RequestHeaders`. I don't think we need to change anything because of this. I'm just noting the similarities and differences so everyone is aware.

dustinsoftware replied:
> || i >= 17 looks super fragile

Agree, I'll rework this

halter73 commented on src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs:
```diff
@@ -79,7 +79,9 @@ public async Task Invoke(HttpContext context)
-        var elements = new string[_fieldsLength];
+        var additionalHeadersLength = _options.CurrentValue.AdditionalRequestHeaders?.Count ?? 0;
+
+        var elements = new string[_fieldsLength + additionalHeadersLength];
```
We should do something to ensure that we clone the set of request headers once we start using it. And we also need to make sure the same cloned set is used in `W3CLoggerProcessor`. Just because it's an `IReadOnlyList<string>` to us doesn't mean it's not an easily mutable `List<string>` to whoever assigned it.

If someone added a header at a bad time, the W3CLoggingMiddleware could throw an `IndexOutOfRangeException`. And even if we avoided that, the headers for the W3C log file would be wrong.

wtgodbe replied:
This reminds me, `FileLoggerProcessor` moves to a new file when the `LoggingFields` change: https://github.com/dotnet/aspnetcore/blob/5ff6e316a8e34bb06b986ec13ec7aae3f61e886b/src/Middleware/HttpLogging/src/FileLoggerProcessor.cs#L64-L74

This PR should do the same when `AdditionalRequestHeaders` changes

dustinsoftware replied:
The headers list is now cloned.

Kahbazi commented on src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs:
```diff
@@ -79,7 +79,9 @@ public async Task Invoke(HttpContext context)
         var options = _options.CurrentValue;
 
-        var elements = new string[_fieldsLength];
+        var additionalHeadersLength = _options.CurrentValue.AdditionalRequestHeaders?.Count ?? 0;
```
```suggestion
        var additionalHeadersLength = options.AdditionalRequestHeaders?.Count ?? 0;
```

Kahbazi commented on src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs:
```diff
@@ -181,6 +183,17 @@ public async Task Invoke(HttpContext context)
                     }
                 }
+
+                if (_options.CurrentValue.AdditionalRequestHeaders != null)
```
```suggestion
                if (options.AdditionalRequestHeaders != null)
```

Kahbazi commented on src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs:
```diff
@@ -181,6 +183,17 @@ public async Task Invoke(HttpContext context)
+                {
+                    for (var i = 0; i < additionalHeadersLength; i++)
+                    {
+                        if (headers.TryGetValue(_options.CurrentValue.AdditionalRequestHeaders[i], out var headerValue))
```
```suggestion
                        if (headers.TryGetValue(options.AdditionalRequestHeaders[i], out var headerValue))
```

dustinsoftware commented:
Ok. Saw the review, should be able to get to it in a few days. :)

On Tue, May 24, 2022 at 12:16, William Godbe ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs
> <https://github.com/dotnet/aspnetcore/pull/41753#discussion_r880712146>:
>
> > @@ -79,7 +79,9 @@ public async Task Invoke(HttpContext context)
>      {
>          var options = _options.CurrentValue;
>
> -        var elements = new string[_fieldsLength];
> +        var additionalHeadersLength = _options.CurrentValue.AdditionalRequestHeaders?.Count ?? 0;
> +
> +        var elements = new string[_fieldsLength + additionalHeadersLength];
>
> This reminds me, FileLoggerProcessor moves to a new file when the
> LoggingFields change:
> https://github.com/dotnet/aspnetcore/blob/5ff6e316a8e34bb06b986ec13ec7aae3f61e886b/src/Middleware/HttpLogging/src/FileLoggerProcessor.cs#L64-L74
>
> This PR should do the same when AdditionalRequestHeaders changes
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/aspnetcore/pull/41753#discussion_r880712146>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHGCFQ4FLZLV6Q4OINEI2LVLT6FFANCNFSM5WMA7RLQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Commit:
Use ISet<string>

Commit:
Add test for invalid header name

Commit:
Add documentation for privacy warnings of logging headers

Commit:
Remove public setter for AdditionalRequestHeaders

dustinsoftware commented:
One potential issue I encountered when testing this out - the order of the fields does matter when printing to the log, however the docs for HashSet indicate that using IEnumerable will yield the elements in no particular order.

https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-6.0

>  A set is a collection that contains no duplicate elements, and whose elements are in no particular order.

In that case, can we use IList / IReadOnlyList instead or some other ordered structure, and dedupe the elements when cloning the list?

halter73 commented:
> In that case, can we use IList / IReadOnlyList instead or some other ordered structure, and dedupe the elements when cloning the list?

We could. That'd require updating the API proposal and going back through the API review process.

However, you don't control the order of the `W3CLoggingFields` either. While the HashSet itself doesn't provide ordering guarantees, nothing is stopping us from cloning the set into a list sorted alphabetically. I think we should do that instead.

dustinsoftware commented:
Spoke to soon, I think `SortedSet` will work. https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.sortedset-1?view=net-6.0

Commit:
Clone and filter AdditionalRequestHeaders

Commit:
Roll files when AdditionalHeaders changes

Commit:
Fix tests

dustinsoftware commented:
@wtgodbe can you clarify under what cases we can expect options to be reloaded? Will `IOptionsMonitor` detect if `AdditionalHeaders.Add` is called from application code? Does this middleware also support watching changed config files from disk?

Commit:
Use separate array for additional headers

dustinsoftware commented:
At this point I believe I have addressed all the feedback except for how options refreshing should work. 

It _looks_ like _loggingFields does not update if the options are reloaded already, since the ctor for W3CLoggerProcessor copies the value.. 

https://github.com/dotnet/aspnetcore/blob/9ea3931a335a55bdcc4fd25fb8a95f02459fe3a8/src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs#L21

halter73 commented:
> @wtgodbe can you clarify under what cases we can expect options to be reloaded? Will `IOptionsMonitor` detect if `AdditionalHeaders.Add` is called from application code? Does this middleware also support watching changed config files from disk? 

`IOptionsMonitor` does not detect changes made directly to the `W3CLoggerOptions` object. The easiest way to update options in an observable way is to bind options from configuration.


https://github.com/dotnet/aspnetcore/compare/halter73/update-fields-directive?expand=1

dustinsoftware commented:
Thanks, let me know if there's anything further for me to update.

wtgodbe reviewed (approved):
Looks good, thanks! We'll do the header refreshing fix in a separate PR
